sdk/scheduler/src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java

```diff
@@ -168,54 +167,54 @@ private Plan createPlan(List<Step> steps) {
                     .collect(Collectors.toList())
                     .size();
 
-            Integer expectedRunningCount = podInstance.getPod().getTasks().stream()
+            if (dirtyAssets.contains(podInstance.getName())) {
+                logger.info("Pod: {} has been dirtied by another plan, cannot recover at this time.",
+                        podInstance.getName());
+                continue;
+            }
+
+            List<TaskSpec> expectedRunningTasks = podInstance.getPod().getTasks().stream()
                     .filter(taskSpec -> taskSpec.getGoal().equals(TaskSpec.GoalState.RUNNING))
-                    .collect(Collectors.toList())
-                    .size();
+                    .collect(Collectors.toList());
 
-            Predicate<Protos.TaskInfo> isPodRecoverable = t -> {
-                return !dirtyAssets.contains(t.getName());
-            };
+            Integer expectedRunningCount = expectedRunningTasks.size();
 
             logger.info(
                     "Attempting to recover pod tasks: {}",
                     failedTasks.stream().map(taskInfo -> taskInfo.getName()).collect(Collectors.toList()));
 
-            if (!Objects.equals(failedRunningTaskCount, expectedRunningCount)
-                    || !failedTasks.stream().allMatch(isPodRecoverable)) {
+            if (!Objects.equals(failedRunningTaskCount, expectedRunningCount)) {
                 logger.warn("Pod: '{}' is not recoverable. Failed task count: {}, Expected task count: {}",
                         podInstance.getName(), failedRunningTaskCount, expectedRunningCount);
                 continue;
             }
 
-            try {
-                if (failedTasks.stream().allMatch(isPodPermanentlyFailed)) {
-                    logger.info("Recovering permanently failed pod: '{}'", podInstance.getName());
-                    recoveryRequirements.addAll(
-                            recoveryReqProvider.getPermanentRecoveryRequirements(failedTasks));
-                } else if (failedTasks.stream().noneMatch(isPodPermanentlyFailed)) {
-                    logger.info("Recovering transiently failed pod: '{}'", podInstance.getName());
-                    recoveryRequirements.addAll(
-                            recoveryReqProvider.getTransientRecoveryRequirements(failedTasks));
-                }
-            } catch (InvalidRequirementException e) {
-                logger.error("Failed to generate recovery requirement for pod: '{}'", podInstance.getName(), e);
-                continue;
+            Collection<String> tasksToLaunch = expectedRunningTasks.stream()
+                    .map(taskSpec -> taskSpec.getName())
+                    .collect(Collectors.toList());
+
+            if (failedTasks.stream().allMatch(isPodPermanentlyFailed)) {
```

Comment:
Could give more context here via comment and/or log about what the difference between these two is.

is it transient=restart, permanent=replace? If so, the logs could just say "Restarting", "Replacing" to be more explicit than "Recovering". From the description in RecoveryType this appears to be the case.